src/main/java/org/broadinstitute/consent/http/service/UserService.java

```diff
@@ -454,12 +454,135 @@ public void hasValidActiveERACredentials(User user) {
     }
   }
 
+  /**
+   * Compliance method defined in
+   * <a href="https://broadworkbench.atlassian.net/browse/DT-1607">this ticket</a> that implements
+   * a set of rules in order to ensure Library Card and Institution matching rules are
+   * adhered to when authorizing users of the system.
+   * @param user the User being evaluated
+   * @return user with the Institution and Library Card rules applied
+   */
+  public User enforceInstitutionAndLibraryCardRules(User user) {
+    Institution institutionFromEmail = institutionService.findInstitutionForEmail(user.getEmail());
+    Institution institutionFromDatabase = null;
+    try {
+      institutionFromDatabase = institutionService.findInstitutionById(user.getInstitutionId());
+    } catch (NotFoundException nfe) {
+      // do nothing.
+    }
+
+    boolean hasInstitutionMatchingEmailDomain =
+        hasInstitutionMatchingEmailDomain(institutionFromEmail);
+
+    boolean modifiedUser = false;
+
+    if (hasInstitutionMatchingEmailDomain) {
+      if (handleUserWithInstitutionInMap(user, institutionFromEmail, institutionFromDatabase)) {
+        modifiedUser = true;
+      }
+    } else {
+      if (handleUserWithoutInstitutionInMap(user, institutionFromDatabase)) {
+        modifiedUser = true;
+      }
+    }
+
+    if (modifiedUser) {
+      return findUserByEmail(user.getEmail());
+    } else {
+      return user;
+    }
+  }
+
+  @VisibleForTesting
+  protected boolean handleUserWithInstitutionInMap(User user, Institution institutionFromEmail, Institution institutionFromDatabase) {
+    boolean modifiedUser = false;
+    if (!hasMatchingInstitutionInDatabase(institutionFromEmail, institutionFromDatabase)) {
```

Comment:
I don't think a method call is needed here. You can replace this with `equals()`
```suggestion
    if (!institutionFromEmail.equals(institutionFromDatabase)) {
```

Reply:
Can we always expect institutionFromEmail to not be null?  I thought we could have visitors to DUOS who were not users and did not have institutions, but went through OAuth.  @rushtong Am I remembering that correctly?

Reply:
`institutionFromEmail` can indeed be null. Imagine a user from a new institution that we don't know about. That user signs in with `user@new.institution.biz` and we know nothing about it. That user would have successfully auth-ed and has a real identity.